dask_image/skexposure/histogram_matching.py

```diff
@@ -0,0 +1,97 @@
+import dask
+import dask.array as da
+import numpy as np
+
+# Only these types are allowed, since some intermediate memory storage scales
+# as the number of unique values in the range of the dtype.
+DTYPES = (np.uint8, np.uint16, np.int8, np.int16)
+
+
+def _match_cumulative_cdf(source, template):
+    """
+    Return modified source array so that the cumulative density function of
+    its values matches the cumulative density function of the template.
+    """
+    if ((source.dtype not in DTYPES) or (template.dtype not in DTYPES)):
+        raise ValueError('Parallelization of match_histograms implemented '
+                         'only for small integer types, not: %s %s' %
+                         (source.dtype, template.dtype))
+
+    _, src_unique_indices, src_counts = da.unique(source.ravel(),
+                                                  return_inverse=True,
+                                                  return_counts=True)
+    tmpl_values, tmpl_counts = da.unique(template.ravel(), return_counts=True)
+
+    # calculate normalized quantiles for each array
+    src_quantiles = da.cumsum(src_counts) / source.size
+    tmpl_quantiles = da.cumsum(tmpl_counts) / template.size
+
+    # The interpolation is a bottleneck and must be done on local node.
+    # This requires an in-memory array which could have a length equal in size
+    # to the unique values of the underlying dtype.  This is why we limit to
+    # small integers.
+    interp_a_values = dask.delayed(_interpolate)(src_quantiles, tmpl_quantiles, tmpl_values)
+    interp_a_values = da.from_delayed(interp_a_values, dtype=np.float, shape=src_quantiles.shape)
+
+    result = src_unique_indices.map_blocks(lambda chunk, values: values[chunk],
+                                           interp_a_values,
+                                           dtype=interp_a_values.dtype)
+    return result.reshape(source.shape)
+
+
+def _interpolate(src_quantiles, tmpl_quantiles, tmpl_values):
+    src_quantiles, tmpl_quantiles, tmpl_values = dask.compute(src_quantiles,
+                                                              tmpl_quantiles,
+                                                              tmpl_values)
+    return np.interp(src_quantiles, tmpl_quantiles, tmpl_values)
```

Comment:
This returns a float dtype, regardless of what the input data type was. Ideally we would preserve the dtype through the whole computation.

You might try using numpy's `astype()` method to specify explicitly.

Reply:
As mentioned prior, this is the behavior of the original skimage implementation.  In my code, I do take the output and do a dtype conversion as you specify.  I'm hesitant to change the behavior, though, as I'm not familiar with all the use cases.